File: sdks/java/reactor-sdk/src/main/java/inc/reactor/sdk/Stats.java
package inc.reactor.sdk;

import java.util.Optional;

/**
 * A reading of the connection.
 *
 * <p>Held as the JSON the platform sent rather than as a fixed set of fields. What a runtime
 * reports grows over time, and a record that named today's fields would drop tomorrow's silently —
 * so the whole payload is here, with accessors for the ones that are stable.
 *
 * @param raw everything the platform reported
 */
public record Stats(JsonValue raw) {

    /**
     * @param name the measurement
     * @return its value, or empty when this reading does not carry it
     */
    public Optional<Double> number(String name) {
        return raw instanceof JsonValue.JsonObject object ? object.getNumber(name) : Optional.empty();
    }

    /**
     * @param name the measurement
     * @return its value, or empty when this reading does not carry it
     */
    public Optional<String> text(String name) {
        return raw instanceof JsonValue.JsonObject object ? object.getString(name) : Optional.empty();
    }
}

File: sdks/android/reactor-sdk-android/src/main/kotlin/inc/reactor/sdk/android/Stats.kt
package inc.reactor.sdk.android

/**
 * A connection statistics snapshot, as the platform reported it.
 *
 * Deliberately not a typed struct of the fields we happen to know today. The measurement set is
 * the platform's to extend, and a binding that mapped it into fixed fields would silently drop
 * every new one — so this keeps the parsed shape and lets a caller read what is there.
 */
public class Stats internal constructor(
    /** Every measurement, as parsed JSON. */
    public val raw: Map<String, Any?>,
) {
    /**
     * A numeric measurement by key, or null when absent.
     *
     * Returns [Double] rather than a narrower type: JSON numbers arrive as doubles, and counters
     * here are signed — coercing to an unsigned or 32-bit type is how a counter that has wrapped
     * or gone negative becomes a wrong answer rather than a visible one.
     */
    public fun number(key: String): Double? = raw[key] as? Double

    public fun string(key: String): String? = raw[key] as? String

    override fun toString(): String = "Stats(${raw.keys.sorted().joinToString(", ")})"
}
